### PR TITLE
ui: Add Consul API Gateway as an external source

### PR DESCRIPTION
**Pull request by kaxcode**
### ✨ Description:
This PR adds support of Consul API Gateway as an external source.

Support appears in the following areas:
- Service List row items
- Top of Service show page
- Service Topology tab - metrics and `Configure metrics dashboard` messaging hidden
- Service Instances tab - list row items
- Service Intentions tab - actions button hidden
- Top of Service Instance show page

### 📸 Demo:
consul-api-gateway

### ⚡ Backend Changes:
BE Support in an internal repo. The BE support has been merged.

### 🤡 Updates to mock-api:
Updated every endpoint that had `"external-source": "${fake.helpers.randomize(['kubernetes'])}"`.

### 🧪 Testing:
Added an acceptance test for hiding the metrics for a Consul API Gateway.

### 📝 **PR Tasks:**
- [x]  Changelog

### CODE REVIEW

File: ui/packages/consul-ui/app/components/consul/external-source/index.hbs
Comment (johncowen) [inline ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: I see what you were thinking here copying the the `@withInfo` attribute from Consul::Kind. I think that now we have this kind of thing in multiple places it might be worthwhile changing things slightly with the multiplicity in mind.

One thing that I'm not 100% sure of here is having the decision on whether to show the info in two places. One place is here in the component (does external-source === 'consul-api-gateway'?), and the other place is where ever you use the component with the `@withInfo` argument (which needs to be true). Both need to be true to show the info and the info is very specific to external-source being `consul-api-gateway`. I think I follow why you need 2 'knobs' for this, but it would be good to chat through which way to do it. An additional side-note here is that if we are going to use i18n here we should use it 100% for every piece of copy, not just the odd word.

Considering all of the above, I think it would be good to think about this a little more and possibly do it differently using named slots (possibly similar thinking to evan's comment). So I'd suggest splitting this bit off into a separate PR so we can get the initial functionality merged in separately and discuss this `withInfo` stuff in a bit more detail. Lemme know what you think.

Comment (evrowe) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: > So I'd suggest splitting this bit off into a separate PR so we can get the initial functionality merged in separately and discuss this withInfo stuff in a bit more detail. Lemme know what you think.

Would it be alright to merge things more or less as-is and make a point to take a pass at `withInfo`/named slots as a separate issue, knowing that we'd need to come back and update this later? The upshot is that allows the implementation discussion around how to standardize and/or make it easier to use consistent patterns to not be under time pressure, and I'd prefer not to see this blocked on that discussion unless the conversation around what to change and implementation of said changes can be accomplished within a day or two.

Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: I would guess that it could be accomplished within a day or two, but happy to merge as is if need be. It also fits in with your comment here on componentizing this a little more. My main worry would be for it to get left and then spread with the next copypaste and then it suddenly becomes a pattern by accident.

We could put this in a separate PR, merge this one pretty much immediately (yay!) and then work on this almost separate feature over the next day or so, I can't see it missing a release for spending a tiny bit more time on it. I think when I originally wrote this up in as a task we split it potentially into two tasks anyway.

I'm easy either way.

I think maybe at the very least, if we are going to use i18n we shouldn't just i18n separate words and hardcode the rest.

Oh actually, just noticed, we this extra functionality is pretty much just to provide links to docs for more info, and those links don't exist yet, so we're gonna have to come back and amend this anyway. Feels like we could just do that in one hit rather than risk this going out in a release with empty links?

Anyway, as I said I'm easy either way but obvs would prefer to split this off if it was me.

File: ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs
Comment (kaxcode) [reply ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs:68]: Support for intentions will be merged today/tomorrow.

File: ui/packages/consul-ui/mock-api/v1/connect/_
Comment (kaxcode) [reply ui/packages/consul-ui/mock-api/v1/connect/_:99]: Support for intentions will be merged today/tomorrow.

File: ui/packages/consul-ui/app/components/consul/external-source/index.hbs
Comment (evrowe) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: I think that if we as a group agree to circle back and address the standardization/componentizing issue as a fast follow, that's a strong enough commitment to make me comfortable with merging this as-is.

> My main worry would be for it to get left and then spread with the next copypaste and then it suddenly becomes a pattern by accident.

I am happy to prioritize addressing this specific issue so that it doesn't fall into being a pattern. And even if it did, we can change it/reverse it.

> I can't see it missing a release for spending a tiny bit more time on it

I'm not concerned about it missing the release; I'd like us to get into the practice of reducing the turnaround time on reviews/approvals/merges for PRs. I see this particular issue as one that isn't crucial to fix for this PR but that is important that we address separately and quickly, and we can add changes to other parts of the code that could benefit from an update as part of that work as well.

> Feels like we could just do that in one hit rather than risk this going out in a release with empty links?

Is that something we've had an issue with in the past? As long as this isn't being backported to anything earlier than 1.11 I don't expect that would be a problem as docs would be required for the release to go out

Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: > We could put this in a separate PR, merge this one pretty much immediately (yay!) and then work on this almost separate feature over the next day or so

Cool yeah I think we want the same outcome here, just different ways of going about it 😆 

Do we want to address the i18n stuff here or later? Again, I'm easy either way.

---

> Is that something we've had an issue with in the past?

Kind of yeah, just links not being ready and we either leave them or put placeholders in, and then they get left as they are forgotten about, and we end up with links going to 404s or the wrong page (🤔 another thing proper e2e testing could help with)

All in all, nothing here that is a specific blocker for merging as is, as long as we are coming back to anything that isn't addressed here.

File: ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs
Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs:68]: Oh wow, TIL, so now I'm trying to understand _why_ we are hiding the actions dropdown? Should I still be allowed to click on the row (I can currently). What about the actions at the bottom of the form once I click into it? Should I be able to delete and save from there if it's a consul-api-gateway intention?

Should we be showing a badge/pill in the row to show that the intention has been added/is managed via consul-api-gateway? At the moment I just don't see an actions button but with no clue as to why I don't.

File: ui/packages/consul-ui/app/components/popover-menu/layout.scss
Comment (johncowen) [reply ui/packages/consul-ui/app/components/popover-menu/layout.scss:7]: Before:

Screenshot 2021-11-03 at 11 18 21


After:

Screenshot 2021-11-03 at 11 15 09

This change makes all our menus dynamically sized to the containing text, so it essentially shrinks the widths of them all, and I don't think that's what is desired (I could be wrong?). The hardcoded width here will have been taken from the original pixel value from the original designs.

If we take the time to un-hardcode this, this should be a min width I think (we also do a weird fit-content thing somewhere that was for something similar to this, not sure this is the right tool here but thought I'd mention incase). Again, separate PR would be better, but I'm easy.

(also e2e testing/screen-component-grabbing 💯 )

File: ui/packages/consul-ui/app/components/topology-metrics/index.hbs
Comment (johncowen) [reply ui/packages/consul-ui/app/components/topology-metrics/index.hbs:49]: Would it be possible to change this to use `ember-can` so `can("view dashboard for service" item=service)` or similar? ty!

File: ui/packages/consul-ui/mock-api/v1/connect/_
Comment (johncowen) [reply ui/packages/consul-ui/mock-api/v1/connect/_:99]: Oh wow, TIL

File: ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs
Comment (kaxcode) [reply ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs:68]: The _why_ is mentioned in my initial comment. Please read and remember this is temporary. 

There are no create/delete actions. Please refer to the mock-ups. 

I'll ask the designer about the badge/pill.

Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/intention/list/table/index.hbs:68]: Cool gotcha

File: ui/packages/consul-ui/app/components/topology-metrics/index.hbs
Comment (kaxcode) [reply ui/packages/consul-ui/app/components/topology-metrics/index.hbs:49]: No, this conditional works. Using `ember-can` for this specific case will open this PR to other changes that will be out of scope.

File: ui/packages/consul-ui/app/components/popover-menu/layout.scss
Comment (kaxcode) [reply ui/packages/consul-ui/app/components/popover-menu/layout.scss:7]: I've added back `min-width`.

File: ui/packages/consul-ui/app/components/consul/external-source/index.hbs
Comment (evrowe) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: I think this looks good to merge. We have some follow-up items we've identified that can be addressed separately once this lands and that we can get done in time for the release.

> Do we want to address the i18n stuff here or later?

I agree that if we are going to be using i18n/localization in the app that we need to be using it more thoroughly, however I'd like there to be a more formal discussion and decision process (as part of general team working agreements) around this before we start requiring for every part of a PR. Let's leave as-is for now but it's on my list of things for us to discuss as a group in the near future.

Comment (johncowen) [reply ui/packages/consul-ui/app/components/consul/external-source/index.hbs:44]: Sounds good

File: ui/packages/consul-ui/app/components/topology-metrics/index.hbs
Comment (johncowen) [reply ui/packages/consul-ui/app/components/topology-metrics/index.hbs:49]: Realised we might need to hide the `Configure dashboard` messaging in the other place that we show this button.

Q: Do you know what `Kind`s could have the metadata `consul-api-gateway`?

**Review by johncowen — APPROVED** [review]
Cool, I think this is good to go. I left a couple more comments/questions, but the only one that I think might be actionable in this PR is the one re: the other `Configure dashboard` button. Not sure if you want to do that here or in another PR, but I'll leave an approval here anyway so its up to you.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/500307.

**Pull request merged**
